packages/wrangler/src/tail.ts

```diff
@@ -0,0 +1,255 @@
+function translateCliFiltersToApiFilterMessage(
+  cliFilters: CliFilters
+): ApiFilterMessage {
+  const apiFilters: ApiFilter[] = [];
+
+  // TODO: do these all need to be their own functions or should
+  // they just be inlined
+  if (cliFilters.samplingRate) {
+    apiFilters.push(parseSamplingRate(cliFilters.samplingRate));
+  }
+
+  if (cliFilters.status) {
+    apiFilters.push(parseOutcome(cliFilters.status));
+  }
+
+  if (cliFilters.method) {
+    apiFilters.push(parseMethod(cliFilters.method));
+  }
+
+  if (cliFilters.header) {
+    apiFilters.push(parseHeader(cliFilters.header));
+  }
+
+  if (cliFilters.clientIp) {
+    apiFilters.push(parseClientIp(cliFilters.clientIp));
+  }
+
+  if (cliFilters.search) {
+    apiFilters.push(parseQuery(cliFilters.search));
+  }
+
+  return {
+    filters: apiFilters,
+    // if debug is set to true, then all logs will be sent through.
+    // logs that _would_ have been blocked will result with a message
+    // telling you what filter would have rejected it
+    debug: false,
+  };
+}
+
+function parseSamplingRate(samplingRate: number): SamplingRateFilter {
+  if (samplingRate <= 0 || samplingRate >= 1) {
+    throw new Error("A sampling rate must be between 0 and 1");
+  }
+
+  return { sampling_rate: samplingRate };
+}
+
+function parseOutcome(
+  statuses: Array<"ok" | "error" | "canceled">
+): OutcomeFilter {
+  const outcomes: string[] = [];
+  for (const status in statuses) {
+    switch (status) {
+      case "ok":
+        outcomes.push("ok");
+        break;
+      case "canceled":
+        outcomes.push("canceled");
+        break;
+      // there's more than one way to error
+      case "error":
+        outcomes.push("exception");
+        outcomes.push("exceededCpu");
+        outcomes.push("unknown");
+        break;
+      default:
+        break;
+    }
+  }
+
+  // filter for unique values
+  const uniqueOutcomes = outcomes.filter(
+    (value, i, arr) => arr.indexOf(value) === i
+  );
+
+  return {
+    outcome: uniqueOutcomes,
+  };
+}
+
+// we actually don't need to do anything here
+function parseMethod(method: string[]): MethodFilter {
+  return { method };
+}
+
+function parseHeader(header: string): HeaderFilter {
+  // headers of the form "HEADER-KEY: VALUE" get split.
+  // the query is optional
+  const [headerKey, headerQuery] = header.split(":", 2);
+  return {
+    header: {
+      key: headerKey.trim(),
+      query: headerQuery?.trim(),
+    },
+  };
+}
+
+// TODO: we _could_ validate this if we wanted to but seems extraneous
+function parseClientIp(ips: string[]): ClientIpFilter {
+  return { client_ip: ips };
+}
+
+function parseQuery(search: string): QueryFilter {
+  return { query: search };
+}
+
+export function prettyPrintLogs(_data: WebSocket.RawData): void {
+  throw new Error("TODO!");
+}
+
+export function jsonPrintLogs(data: WebSocket.RawData): void {
+  console.log(JSON.stringify(JSON.parse(data.toString()), null, "  "));
```

Comment:
I prefer we avoid strings of spaces, since it is not so easy to count the number of chars.

```suggestion
  console.log(JSON.stringify(JSON.parse(data.toString()), null, 2));
```